### ISSUE TITLE
Problems Installing GMMAT package on Windows

### ISSUE THREAD

**Issue by ngwali** [hanchenphd/GMMAT#2]
Dear Han,

 Sorry for disturbing you sir.  Could you please help me decipher why the GMMAT package package does not completely  install on my Windows PC?. I am using the latest version of R  and RStudio.

I tried using MAC and Windows, but failed. Below is a warning message after several tries on Windows.

 **Successfully installed all dependencies:**
>install.packages(c("devtools", "RcppArmadillo", "CompQuadForm", "foreach", "testthat"), repos = "https://cran.r-project.org/")
>source("https://bioconductor.org/biocLite.R")
>biocLite(c("SeqArray", "SeqVarTools"))

**Unsuccessful installation of GMMAT**

> devtools::install_github("hanchenphd/GMMAT")
Downloading GitHub repo hanchenphd/GMMAT@master
Skipping 2 packages not available: SeqArray, SeqVarTools
√  checking for file 'C:\Users\Public\Documents\Wondershare\CreatorTemp\Rtmp6bFliw\remotes45dc74f22743\hanchenphd-GMMAT-e5a0967/DESCRIPTION' ... 
-  preparing 'GMMAT':
  √  checking DESCRIPTION meta-information ... 
-  cleaning src
√  checking vignette meta-information ...
-  checking for LF line-endings in source and make files and shell scripts
-  checking for empty or unneeded directories
-  looking to see if a 'data/datalist' file should be added
-  building 'GMMAT_1.0.3.tar.gz'

Installing package into ‘C:/Users/LENOVO/Documents/R/win-library/3.5’
(as ‘lib’ is unspecified)
* installing *source* package 'GMMAT' ...
** libs

*** arch - i386
C:/RBuildTools/3.5/mingw_32/bin/g++  -I"C:/PROGRA~1/R/R-35~1.2/include" -DNDEBUG  -I"C:/Users/LENOVO/Documents/R/win-library/3.5/Rcpp/include" -I"C:/Users/LENOVO/Documents/R/win-library/3.5/RcppArmadillo/include"        -O2 -Wall  -mtune=generic -c fitglmm.cpp -o fitglmm.o
fitglmm.cpp:51:19: fatal error: bzlib.h: No such file or directory
#include <bzlib.h>
^
  compilation terminated.
make: *** [C:/PROGRA~1/R/R-35~1.2/etc/i386/Makeconf:215: fitglmm.o] Error 1
ERROR: compilation failed for package 'GMMAT'
* removing 'C:/Users/LENOVO/Documents/R/win-library/3.5/GMMAT'
* restoring previous 'C:/Users/LENOVO/Documents/R/win-library/3.5/GMMAT'
In R CMD INSTALL
Error in i.p(...) : 
  (converted from warning) installation of package ‘C:/Users/Public/Documents/Wondershare/CreatorTemp/Rtmp6bFliw/file45dc568b5e47/GMMAT_1.0.3.tar.gz’ had non-zero exit status

Thank you very much and I await your response. I also sent you a similar question via your email. 

Thanks.

**Comment by hanchenphd** [comment]
It means your Windows toolset could not find the header file for bzip2. If you don't have it, you can download a binary distribution of external software (see [https://cran.r-project.org/doc/manuals/r-devel/R-admin.html#Getting-the-tools](url)). Then in your Makeconf file, specify the location of external software as LOCAL_SOFT.

**Comment by ngwali** [comment]
Hell Han;

              Thanks for getting back to me on such a short notice.

1) I checked the Rtools and it is installed as RBuildTools in **C:\RBuildTools**
2) I noticed that Rstudio does not recognize "RBuildTools" ,  but rather R itself uses Rtools. So I had to install Rtools maunally. Thus in my "C:\" there are now both RBuildTools and Rtools now installed, such that when I tried installing GMMAT using main R command line, i recognizes Rtools and when using R-Studio, it recognizes RBuildTools respectively.

3) I opened the **Makeconf** file to specify the location of external software as **LOCAL_SOFT**, but found out that **LOCAL_SOFT** was already there. Please can you take a look at the **Makeconf** file attached below?

[Makeconf.txt](https://github.com/hanchenphd/GMMAT/files/2802465/Makeconf.txt)

4) I think that R is trying to install a tarbool file of the GMMAT package, which is used on MAC rather than on windows. e.g -building 'GMMAT_1.0.3.tar.gz' from github.

Please can take a look at the Makeconf file and see if what you meant was solved already, because I still get thesame error message.

5) If need be, can you please configure the file and send back to me? Otherwise, if you have the '**GMMAT_1.0.3.zip**' rather than the tarball file '**GMMAT_1.0.3.tar.gz**', I could install manually and dont have to download from github directly.

NB: I am not really that good with building R packages, but simple know how to apply and use them. Thus I give you applause for this, and hope that you help me out figure this issue.

Thank you professor.

Best regards

**Comment by ngwali** [comment]
Hello Han;

                Since I could not get the GMMAT package Installed in R, I decided to use the written functions of the package, load the functions using **source()**  and including the the **example** data,  and then attempted to  reproduce your examples in the GMMAT manual. 

Unfortunately, the program failed due to the following errors.

**C_fitglmm.ai  not found**

See code below.

> source("C:\\Users\\LENOVO\\Downloads\\GMMAT-master\\R\\glmm.score.r")
> source("C:\\Users\\LENOVO\\Downloads\\GMMAT-master\\R\\glmm.wald.r")
> source("C:\\Users\\LENOVO\\Downloads\\GMMAT-master\\R\\glmmkin.r")
> source("C:\\Users\\LENOVO\\Downloads\\GMMAT-master\\R\\SMMAT.r")

 **#loading example  data**
> load("C:/Users/LENOVO/Downloads/GMMAT-master/data/example.rda")
> attach(example)_
The following objects are masked from example (pos = 4):

    GRM, pheno, pheno2

**#running the  null model**
> model0 <- glmmkin(disease ~ age + sex, data = pheno, kins = GRM, 
+               id="id"    , family = binomial(link = "logit"))_**

 **Error in glmmkin.ai(fit0 = fit0, kins = kins, covariance.idx = covariance.idx,  : 
  object 'C_fitglmm_ai' not found** 

Notes: I think this should be another alternative to run your GMMAT package, but it still fails?? I can reallyget it to work. I also think that even if I succeed installing the GMMAT package in R, it I will still get thesame problem  object '**C_fitglmm_ai' not found** as I showed above.

Unless otherwise, I will really need your help with this.

Thanks.

**Comment by hanchenphd** [comment]
GMMAT contains C++ code, so simply sourcing its R code without compiling its C++ code is not supposed to work. A binary version for Windows users might be available on CRAN soon, but you have to wait for it patiently because the CRAN folks are really busy checking, building and managing thousands of R packages.

Suppose you have downloaded a binary distribution of external software that contains bzip2, for example, in the folder C:/external_software (or any other path you like), please verify that you have the following files available:
C:/external_software/include/bzlib.h
C:/external_software/lib$(R_ARCH)/libbz2.a

Then go back to your Makeconf file, uncomment the line
\#LOCAL_SOFT
and change it to
LOCAL_SOFT = C:/external_software

After that you will be able to install GMMAT from source on your Windows machine.

**Comment by ngwali** [comment]
Hello Han;

 I appreciate your help with this. But I did not succeed in
 installing the GMMAT package,  However, I tried installing
 the GMMAT_Lite, but the glmmscore function doesn't
 work either. Thank you very much.

However, if you know of any alternative, or have 
the Makeconf file edited and sent to me, I will be happy.

Thanks very much sr.

**Comment by hanchenphd** [comment]
We used to have a version 0.7-1 for people who do not have the C++ library boost on their machines (see Section 8.3 of the [user manual](https://github.com/hanchenphd/GMMAT/blob/master/inst/doc/GMMAT.pdf), but back then we did not support Windows and this version is not supposed to work on Windows machines, and it is no longer updated or maintained.

Our current version 1.0.3 fully supports Windows portability, and it does not require dependency on the C++ library boost any more. I understand that compiling C++ code on Windows from source is no fun, until a binary zip file is posted on CRAN. But did you follow my instructions posted above? Which step are you stuck in?
